### PR TITLE
New seed data

### PR DESCRIPTION
**Pull request by pieterlukasse**
New seed DB.

These files contain the updated seed database for cBioPortal. We included:
1. Updated genes (60,072) and gene aliases
2. Updated gene lengths
3. Updated pfam graphics
4. Updated the `type_of_cancer` table, according to `cgds_public` (now 601 records in total)
5. Updated the `interaction` table, according to `cgds_public` and the new genes list (step 1 above) (now 523,978 records in total)

Compared to the previous seed database, the following number of values are removed because their associated Entrez Gene IDs do not exist anymore in the NCBI Gene info:
- cosmic_mutation: 247
- uniprot_id_mapping: 1375
- interaction: 56
- drug_interaction: 6

These tables were dropped, since they were not found in cgds.sql:
- micro_rna
- micro_rna_alteration
- mutation_frequency
- patient_list
- patient_list_list

Notify:
@sheridancbio @sandertan

### CODE REVIEW

**Comment by pieterlukasse** [comment]
@sheridancbio : here is the PR with data only. Maybe you can test this one with `cgds.sql`? If it works well, I suggest we merge this and make it the first version.

**Comment by pieterlukasse** [comment]
and we would have to update the documentation in #1696

**Comment by sheridancbio** [comment]
sounds good. I will try to get to this today

**Comment by sheridancbio** [comment]
There were changes to the seed database that I recommended in cBioPortal/cbioportal#1696:

Drop table clinical_trial_keywords
Add field IMPORT_DATE in the cancer_study table (to match cgds)
Change the name of index PATIENT_ID to SAMPLE_ID in table sample_list_list
Remove the AUTO_INCREMENT=2 clause on table genetic_profile
Drop default '0' clause for field SAMPLE_ID in table sample_list_list
Drop default '0' clause for field ENTREZ_GENE_ID from table uniprot_id_mapping

Also, there were some new tables added in 1.3.0 that should be present in the seed.

**Comment by pieterlukasse** [comment]
@sheridancbio right, so you won't find most of these in the files I uploaded since they are *data only* files. I think the only part that needs to be removed is the data that gets added to `clinical_trial_keywords` in this case. I can push a new .sql without any insert statements referring to this table. 

The other changes would have to happen in cgds.sql (if not done yet).

**Comment by sheridancbio** [comment]
Ok, great. Also note: the clinical_trials table has been dropped as well (in case there are any data loading statements for that table)

**Comment by pieterlukasse** [comment]
@sheridancbio I pushed up the new version

**Comment by sheridancbio** [comment]
One issue I have found is that the seed database contains the data for the info table (INSERT INTO `info` VALUES ('1.2.1'). However, the cgds script itself sets up the record in info with the version number corresponding to the schema defined in cgds.sql. After running cgds.sql and importing the data in seed-cbioportal_no-pdb_hg19.sql, there were two records (conceivably with different values for the version number) in the info table.

We should either take the insert statement out of cgds, or out of seed-cbioportal

Also, the arrangement of the info table (with the possibility of two records) seems improper to me. I don't know what other kinds of info we might want to put into this table, but I think maybe a key/value pair organization would be better. Maybe like this:
CREATE TABLE `info` (
`INFO_KEY` varchar(256),
`INFO_VALUE` varchar(8096),
PRIMARY KEY (`INFO_KEY`)
);
Then we could do this for setting / updating the value:
DELETE FROM `info` where `INFO_KEY` = 'DB_SCHEMA_VERSION';
INSERT INTO `info` VALUES ('DB_SCHEMA_VERSION','1.2.1');

This would make it impossible to store two database version numbers in the table.

**Comment by pieterlukasse** [comment]
@sheridancbio nice catch! I removed the insert statement from the seed .sql file. I think it should be in cgds.sql. 

I also agree with your suggestion for the key field in the `info` table. Let's do that in cgds.sql and migrate.sql.

**Comment by sheridancbio** [comment]
the new seed database data is schema-compatible with 1.2.1 cgds.sql. Examining data (comparing to cgds_public) there are minor reductions to several tables due to dropped / no longer listed entrez gene ids. But there is a major reduction in size (80% less) for table interaction, and there are 95 fewer records in type_of_cancer. 
Tables with fewer records:
	cosmic_mutation
	drug_interactions
	gene (dropped uniques?)
	gene_alias (65k instead of 73k)
	interaction (153k instead of 524k) ***
	pfam_graphics (21k instead of 23k)
	type_of_cancer (551 instead of 646)
	uniprot_id_mapping
For type_of_cancer, there are many examples of records in cgds_public such as:
oovc    Ovarian Cancer, Other
osaca    Salivary Carcinoma, Other
pns    Peripheral Nervous System
raml    Renal Angiomyolipoma
tnet    Thymic Neuroendocrine Tumor
But also cases of records which alter Name or ID:
-----------------
< lgcos Low Grade Central Osteosarcoma
< lgess Low Grade Endometrial Stromal Sarcoma
---
> lgcos Low-Grade Central Osteosarcoma
> lgess Low-Grade Endometrial Stromal Sarcoma
-----------------
419a479
> raml Renal Angiomyolipoma
423c483,484
< rea Renal Epithelioid Angiomyolipoma

File: seedDB/README.md [hunk 1]
```diff
@@ -1,25 +1,7 @@
-Files to seed a new cBioPortal database.
+These files are MySQL database dump for seeding a new instance of the cBioPortal DB. They contain all the necessary background data for a properly functioning cBioPortal website, including cancer types, gene, uniprot-mappings, drug, pdb and network data. Data is compatible with release **1.3.1** of cBioPortal.
```
Comment (sheridancbio) [inline seedDB/README.md]: I am not sure that the database schema number and the cBioPortal codebase version number are coupled to each other. Already the public release at cbioportal.org reports a 1.3.2 version number, while the cgds.sql script in rc seems to be at 1.3.0 and migration.sql is at 1.3.1 [incidentally I think that is a mistake --- I would argue that anytime migration.sql gets incremented there should be a matching increment to cgds.sql].  But the main point is that the seed db will track a version of the schema -- which will often be different than the codebase version.
I would maybe add an explanation of the two different versions:
- codebase [from pom.xml I think]
- DB_SCHEMA_VERSION [from cgds/migration, and checked against portal.properties db.version

**Comment by sheridancbio** [comment]
I'm taking a look now at the differences in the network table size. I'll ask about the changes here related to that and the cancer type codes.

**Comment by sheridancbio** [comment]
cc: @n1

**Comment by sheridancbio** [comment]
looking at the type_of_cancer record differences, here is a renamed type_of_cancer description:

The proposed new seed database has:
epis Epitheloid Sarcoma
The current cbioportal databases (cgds_public and cgds_gdac) has:
epis Epithelioid Sarcoma

So which is correct? and where did the correction / error come from? @n1zea144 recommended that we should base the type_of_cancer on the current database for the oncotree. If the TYPE_OF_CANCER_ID values do not cover the studies which are provided for importing, then import for those studies may fail and users would need to manually correct the type_of_cancer table, so we should try to update this table to what we currently have. We could dump the type_of_cancer table from the current production databases here .. or better yet, we could extract programmatically the necessary info from the google spreadsheet portal_importer_configuration. (4 levels in the form {name (type_of_cancer_id)})


Sometimes there are missing records for type of cancer. For instance The proposed new seed database does not have the type_of_cancer record:
raml Renal Angiomyolipoma
which *is* present in the current cbioportal databases. Possibly, these additional records get added during study loading.

**Comment by sheridancbio** [comment]
I have dumped the current cgds_public type_of_cancer table. Maybe it would be easier to just use this one for now.

Also, I see that there is a discrepancy between cgds_public and cgds_gdac. cgds_public has 646 records and cgds_cgda has 551.  I'm sending the bigger one by email.

**Comment by sheridancbio** [comment]
I have examined the interaction tables (comparing the new seed db to the cgds_public). There is a large disparity between entrez_gene_ids referenced. There are 3449 genes present in the new seed database which are not present in cgds_public, and 2728 gene references present in cgds_public which are not in the new seed database. I suspect that this large change in gene references is the cause of the 80% reduction in interaction records.

**Comment by sheridancbio** [comment]
Again, perhaps we can use a dump of the current interactions database. cgds_public and cgds_gdac both have the same content : 524395 records. I've dumped them and will email you.

File: seedDB/README.md (hunk 1 above)
Comment (pieterlukasse) [inline seedDB/README.md]: Good point. I updated the text to make the versioning story more explicit. 
Regarding cgds.sql at 1.3.0, this is not a mistake. The migration.sql is also at this same 1.3.0: see https://github.com/cBioPortal/cbioportal/blob/v1.3.1/core/src/main/resources/db/migration.sql

Comment (pieterlukasse) [inline seedDB/README.md]: and db.version in https://github.com/cBioPortal/cbioportal/blob/v1.3.1/pom.xml is also 1.3.0, so this is good.

**Comment by pieterlukasse** [comment]
Hi @sheridancbio , I updated the tables. Here is a log of the actions: 

1- Imported the .sql files you sent. Below are the counts *after* import: 
```
mysql> select count(*) from interaction;
+----------+
| count(*) |
+----------+
|   524395 |
+----------+

mysql> select count(*) from type_of_cancer;
+----------+
| count(*) |
+----------+
|      601 |
+----------+
```


2- Checked if any gene in the `interaction` table was invalid (i.e. not existing anymore in the updated `gene` table):

```
mysql> Select count(*) from interaction where GENE_A is not null and GENE_A not in (Select ENTREZ_GENE_ID from gene);
+----------+
| count(*) |
+----------+
|      178 |
+----------+

mysql> Select count(*) from interaction where GENE_B is not null and GENE_B not in (Select ENTREZ_GENE_ID from gene);
+----------+
| count(*) |
+----------+
|      239 |
+----------+
```
3- Deleted the invalid genes, resulting in slightly less `interaction` records: 
```
mysql> DELETE from interaction where GENE_B is not null and GENE_B not in (Select ENTREZ_GENE_ID from gene);         
Query OK, 239 rows affected (0.56 sec)

mysql> DELETE from interaction where GENE_A is not null and GENE_A not in (Select ENTREZ_GENE_ID from gene);         
Query OK, 178 rows affected (0.48 sec)

mysql> select count(*) from interaction;       
+----------+
| count(*) |
+----------+
|   523978 |
+----------+
```

4 - I pushed up the new seed .sql file (see commit https://github.com/cBioPortal/datahub/pull/13/commits/915697b2ab80aaa907a17fde0232586705ddd1c2)

**Comment by pieterlukasse** [comment]
I also updated the description of this PR to reflect last changes.

File: .gitattributes [hunk 2]
```diff
@@ -1,3 +1,4 @@
 *.tar.gz filter=lfs diff=lfs merge=lfs -text
 *.pdf filter=lfs diff=lfs merge=lfs -text
 *.txt filter=lfs diff=lfs merge=lfs -text
+*.gz filter=lfs diff=lfs merge=lfs -text
```
Comment (sheridancbio) [inline .gitattributes]: Doesn't this line (\*.gz) shadow the first line (\*.tar.gz) ?  We can drop the first line I think.

**Comment by sheridancbio** [comment]
I will merge now. We have addressed the two prominent discrepancy issues in the seed databases. A desirable test would be to build a database from this seed and with importing all prepared studies available through datahub.

We also should strive to automate (or at least document) the procedure for collecting the reference data in the seed database. That way, a static resource which collects these few tables will not be necessary. (possibly connected to the development of https://github.com/cBioPortal/importer)

**Comment by sheridancbio** [comment]
reverting and squashing

**Comment by sheridancbio** [comment]
Never mind .. reverting just adds to the commit history. To truly squash I would need to rewrite history, which I am not willing to do in a public repository. Apologies for the bloat in the timeline.

**Comment by sheridancbio** [comment]
hmm .. I see that this is in the rc branch and not master. Perhaps squashing inside rc is all right. I'm going to try.

**Comment by sheridancbio** [comment]
I have rebased rc to be 2 commits, squashing 13. The two commits have checksums:
b69c86803c40d543080bf31a645721d06c82d08d  (this is 13 commits, the main data)
86249ff58a8f3e657b73664923e9fb634eb51a82  (this is the final commit which sets the download URL in the instructions)

It is important (to avoid re-introducing the squashed changesets) to rewrite the history in the fork repository thehyve on branch new_seed_data to match these two commits. (or that branch can be deleted perhaps).

To rewrite, I think it would be done somewhat like this (warning: not tested) :
(if needed -->)  git remote add upstream https://github.com/cBioPortal/cbioportal.git
(if needed -->)  git remote add origin https://github.com/thehyve/datahub.git
git checkout new_seed_data
git reset --hard babaf21f21430adc8441eabb4447af2cba7b739e
git pull upstream rc
git push origin new_seed_data --force

**Comment by pieterlukasse** [comment]
Hi @sheridancbio , thanks for merging and trying out the squash. 

I think that for this repository it is not ideal to rewrite history. Github does not handle this well and I am always confused by the outcome. For example, github is now reporting that what you committed is something from 23 days ago (see screenshot). So to be sure you made no mistake, I will have to download the files, load them into a DB again and check the contents to see if it contains the last data. The same is true for other (source code) repositories btw...but there it is easier to check if all went well. 

I would therefore like to propose that we do not rewrite history in this repository when it regards the LFS stored data files. Let me know what you think.

Screenshot: 
image

**Comment by sheridancbio** [comment]
In general, I agree. I had mixed feelings about doing this. I'm not afraid that anything went wrong during the squash operation though; my main worry is that rewriting history will cause problems for all developers who had downloaded between the acceptance of the PR and the squash -- for them, their local repository would be out of sync with the github repository and there is the risk / likelihood that the squashed changesets would be re-pushed back to github in a future PR.

The main reason that convinced me to go ahead was that there were several intermediate versions of these large data files, and these "middle" versions were no longer actually useful, and in fact would be harmful if someone choice a point in the timeline when things were erroneous / incomplete / inconsistent. By squashing the intermediate versions, we make the timeline cleaner and with fewer unstable commits. But I should have squashed at the time of PR acceptance --- there is an option for doing that but I forgot to use it.

But maybe you are right. once a PR is merged to a branch in the main github repo, we should just live with the fact that there are transitory commits in the timeline, and rely on users to know the difference between a stable commit and a "work in progress" commit. In my mind, I convinced myself that that rule should apply to branch master .. but maybe all named branches are equal entities in the repo.

**Comment by sheridancbio** [comment]
Talking with @n1zea144 we realize that another easier approach would have been to squash at the time that the rc branch gets merged to master. Here is the menu of choices in github:
githubsquashmenu

In this case we don't want to squash these two commits because the commit checksum for the link must be kept consistent, but this might be an option going forward (squashing during merge to master)

**Pull request merged**